### PR TITLE
add web compatibility

### PR DESCRIPTION
**Pull request by hakanson**
*Issue #, if available:*

*Description of changes:*

- add web specific implementation and use subpath imports to choose implementation at runtime
- web version uses `webcrypto` instead of node `crypto` to verify JWT and import keys
- web version of `fetch` and added base64url decoding method
- added tests using vite and cypress to test "in browser"

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+**JavaScript** library for **verifying** JWTs signed by **Amazon Cognito**, and any **OIDC-compatible IDP** that signs JWTs with **RS256** / **RS384** / **RS512**.
 
 ## Installation
 
 `npm install aws-jwt-verify`
 
-Note: this library can be used with Node.js 14 or higher. If used with TypeScript, TypeScript 4 or higher is required.
+This library can be used with Node.js 14 or higher. If used with TypeScript, TypeScript 4 or higher is required.
+
+This library can also be used in Web browsers (modern browsers that support e.g. `SubtleCrypto`).
```
Comment (leelalagudu) [inline README.md:11]: As discussed , call out major browser versions that are supported (i.e. support `SubtleCrypto`)

File: jest.config.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ module.exports = {
   preset: "ts-jest",
   testEnvironment: "node",
   moduleFileExtensions: ["ts", "js"],
-  collectCoverageFrom: ["src/*.ts"],
+  collectCoverageFrom: ["src/*.ts", "!src/node-web-compat-web.ts"],
```
Comment (leelalagudu) [inline jest.config.js:5]: Is the exclusion because Cypress tests are covering the unit tests for the code here?

Comment (ottokruse) [reply jest.config.js:5]: Yes indeed

File: README.md (hunk 1 above)
Comment (hakanson) [reply README.md:11]: Looking on MDN:
- [SubtleCrypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) importKey and verify were supported in Chrome 37, Edge 12, Firefox 34, and Safari 11
- [TextEncoder](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder) was supported in Chrome 38, Edge 79, Firefox 18, and Safari 10.1
- [fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch) was supported in Chrome 42, Edge 14, Firefox 39, and Safari 10.1

Of these, Edge 79 (the first Chromium version, release in Jan 2020) is the most recent on the list.  In fact, only IE11 wouldn't natively support our library.  I'm tempted just to strike the "modern browser" parenthetical.

Comment (ottokruse) [reply README.md:11]: Fair enough

File: src/assert.ts [hunk 3]
```diff
@@ -14,12 +14,12 @@ import { AssertionErrorConstructor, FailedAssertionError } from "./error.js";
  * @param expected - The expected value
  * @param errorConstructor - Constructor for the concrete error to be thrown
  */
-export function assertStringEquals(
+export function assertStringEquals<T extends string>(
```
Comment (leelalagudu) [inline src/assert.ts:17]: What is the objective of this generic that extends string? If the aim is to enforce strong typing, we could do that through type assertion check inside the function right?

File: tests/vite-app/run-tests.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+run_test() {
+    START_SERVER_CMD=$1
+    SERVER_PORT=$2
+    START_CYPRESS_CMD=$3
+
+    echo "Starting Vite server in background ..."
+    $START_SERVER_CMD &
+    VITE_PID=$!
+    echo "Vite server launched as PID ${VITE_PID}"
+    while ! nc -z localhost $SERVER_PORT; do
```
Comment (leelalagudu) [inline tests/vite-app/run-tests.sh:12]: How do we handle a stuck server boot process , for ex in a port conflict scenario?

File: src/assert.ts (hunk 3 above)
Comment (ottokruse) [reply src/assert.ts:17]: This is pure TypeScript developer experience.

By adding the `asserts actual is T` as function return type, we make the assertions of this function available to TypeScript's control flow analysis, such that calling code will then understand that the type of `expected` is a `string`.

This is explained by TypeScript's lead architect here: https://github.com/microsoft/TypeScript/pull/32695

I added this in because it's simple to add, and makes using these functions a bit nicer in TypeScript.

File: tests/vite-app/run-tests.sh (hunk 4 above)
Comment (ottokruse) [reply tests/vite-app/run-tests.sh:12]: Yes it would run forever until the build times out (or if run by developer until they CTRL-C).
Acceptable since this is just test runner code (?)

Comment (leelalagudu) [reply tests/vite-app/run-tests.sh:12]: I would still prefer if we put in an upper time bound , for ex `while ! nc -z localhost $SERVER_PORT; [[ "$SECONDS" -lt 2000 ]]; do` , assuming that the server boot up should not take more thant 2000 seconds since the shell has been spawned

File: README.md (hunk 1 above)
Comment (leelalagudu) [reply README.md:11]: @hakanson , I assume you're updating the readme, correct?

Comment (hakanson) [reply README.md]: just committed that change

File: tests/vite-app/run-tests.sh (hunk 4 above)
Comment (ottokruse) [reply tests/vite-app/run-tests.sh:13]: Done (capped at 30 seconds, which should be ample)

**Review by leelalagudu — APPROVED** [review]
LGTM

File: README.md [hunk 5]
```diff
@@ -75,6 +77,12 @@ This library was specifically designed to be easy to use in:
 - [CloudFront Lambda@Edge](https://docs.aws.amazon.com/lambda/latest/dg/lambda-edge.html)
 - Node.js APIs, e.g. running in AWS Fargate, that need to verify incoming JWTs
 
+## Usage in the Web browser
+
+Many webdev toolchains (e.g. [CreateReactApp](https://github.com/facebook/create-react-app)) make including `npm` libraries in your web app easy, in which case using this library in your web app should just work.
+
+If you need to bundle this library manually yourself, be aware that this library uses [subpath imports](https://nodejs.org/api/packages.html#subpath-imports), to automatically select the Web crypto implementation when bundling for the browser. This is supported out-of-the-box by [webpack](https://webpack.js.org/) and [esbuild](https://esbuild.github.io/). An example of using this library in a Vite web app, with Cypress tests, is included in this repository [here](tests/vite-app/).
```
Comment (elorzafe) [inline README.md:84]: This is great stuff!

File: tests/unit/jwt-rsa.test.ts [hunk 6]
```diff
@@ -203,6 +204,28 @@ describe("unit tests jwt verifier", () => {
...
+            audience: null,
+            issuer: null,
+          });
+        expect.assertions(2);
+        mockHttpsUri("https://example.com/path/to/jwks.json", {
+          responsePayload: JSON.stringify(keypair.jwks),
+        });
+        await expect(statement).rejects.toThrow("Invalid signature");
+        mockHttpsUri("https://example.com/path/to/jwks.json", {
+          responsePayload: JSON.stringify(keypair.jwks),
```
Comment (elorzafe) [inline tests/unit/jwt-rsa.test.ts:225]: nit: maybe rename `keypair.jwks` to something more descriptive

File: tests/vite-app/cypress/integration/unittests.spec.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      audience: AUDIENCE,
+      jwksUri: JWKSURI,
+    });
+
+    try {
+      const payload = await verifier.verify(EXPIRED_TOKEN);
+
+      expect(payload).to.not.exist;
+    } catch (ex) {
+      expect(ex.message).to.include("Token expired at ");
```
Comment (elorzafe) [inline tests/vite-app/cypress/integration/unittests.spec.ts:44]: nit: You could something like expect to throw. https://docs.cypress.io/guides/references/assertions#BDD-Assertions

**Review by elorzafe — APPROVED** [review]
Just have nit picking comments, nothing blocker.

This is great stuff 🎉

File: src/jwk.ts [hunk 8]
```diff
@@ -171,7 +171,9 @@ export class SimplePenaltyBox implements PenaltyBox {
   registerFailedAttempt(jwksUri: string): void {
     const i = setTimeout(() => {
       this.waitingUris.delete(jwksUri);
-    }, this.waitSeconds * 1000).unref();
+    }, this.waitSeconds * 1000);
+    /* istanbul ignore else */
+    if (i.unref && typeof i.unref === "function") i.unref(); // Node.js only
```
[the remark this answers is not in the document]
Comment (ottokruse) [reply src/jwk.ts:176]: It could it could ...
Could create e.g. a `setTimeoutUnref` function in `node-web-compat` that would automatically `unref()` in the NodeJS implementation. (Can't do that in Web though)
Like that better?
Thought about this one before but then decided not to do it as it's just a small thing and I didn't wanna bother, but I'd be willing to see it your way :)

File: src/jwt-rsa.ts [hunk 9]
```diff
@@ -146,35 +143,41 @@ export type JwtRsaVerifierMultiIssuer<
...
+ * @returns boolean: true if the JWS is valid, or false otherwise
  */
-export enum JwtSignatureAlgorithms {
-  RS256 = "RSA-SHA256",
-  RS384 = "RSA-SHA384",
-  RS512 = "RSA-SHA512",
-}
+export type JwsVerificationFunctionSync = (props: {
+  keyObject: GenericKeyObject;
+  alg: "RS256" | "RS384" | "RS512";
```
[the remark this answers is not in the document]
Comment (ottokruse) [reply src/jwt-rsa.ts:157]: Good idea

File: tests/unit/jwt-rsa.test.ts (hunk 6 above)
Comment (ottokruse) [reply tests/unit/jwt-rsa.test.ts:225]: Like what? To me `keypair.jwks` is descriptive but I'm wondering what your perspective is

File: src/jwt-rsa.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ottokruse) [reply src/jwt-rsa.ts]: Done, let me know if you like this better now

File: tests/vite-app/cypress/integration/unittests.spec.ts (hunk 7 above)
Comment (ottokruse) [reply tests/vite-app/cypress/integration/unittests.spec.ts:44]: @hakanson

Comment (hakanson) [reply tests/vite-app/cypress/integration/unittests.spec.ts:44]: The method under test is an async function, so looking to see if cypress supports [chai-as-promised](https://www.chaijs.com/plugins/chai-as-promised/) assertion language. (e.g. `promise.should.be.rejectedWith(Error)`)

Either way, it still might be better to check the Error type than the message

Comment (hakanson) [reply tests/vite-app/cypress/integration/unittests.spec.ts:53]: no luck getting chai-as-promised working, so added checks like `expect(ex).to.be.an.instanceof(JwtExpiredError)` on the unit tests

Comment (elorzafe) [reply tests/vite-app/cypress/integration/unittests.spec.ts:53]: no worries, is still valid

File: tests/unit/jwt-rsa.test.ts (hunk 6 above)
Comment (elorzafe) [reply tests/unit/jwt-rsa.test.ts:225]: In that case is an invalid keypair right?

File: src/jwt-rsa.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (elorzafe) [reply src/jwt-rsa.ts]: Yeah I like it!

File: src/jwk.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (elorzafe) [reply src/jwk.ts:176]: You dont need to call clearTimeout on Web?

File: tests/vite-app/cypress/integration/unittests.spec.ts (hunk 7 above)
Comment (hakanson) [reply tests/vite-app/cypress/integration/unittests.spec.ts:53]: To get this working, I had to `import { JwtExpiredError } from "aws-jwt-verify/error"` - but now I am wondering with we want to use `index.ts` for barrel imports from `error.ts`?  Something like `export * from "./error"` in that file so that  `import { JwtRsaVerifier, JwtExpiredError } from "aws-jwt-verify" works as expected.

But, anything we add to `index.ts` becomes part of the API contract

Comment (ottokruse) [reply tests/vite-app/cypress/integration/unittests.spec.ts:53]: This is a design choice we made (and you can surely have opinions about):

The main `index.ts` file has just 2 exports, so the namespace is nicely small, which is nice for regular usage - because IDEs will not swamp users with symbols when they enable code completion on the import.

Errors, but also other things you might want to refer to such as from `jwt-model`or `jwk` must be imported as you did from the subpackage.

`import { JwtExpiredError } from "aws-jwt-verify/error"`
`import { JwtPayload } from "aws-jwt-verify/jwt-model"`

We like it like that. Let me rephrase, I like that :) (I think, I went through lengths to get this to work nicely, that's why the package.json has the `exports` object filled with all the submodules, to support this)

Whether you like it too or not -- it __is__ documented explicitly: https://github.com/awslabs/aws-jwt-verify#verification-errors

File: tests/unit/jwt-rsa.test.ts (hunk 6 above)
Comment (ottokruse) [reply tests/unit/jwt-rsa.test.ts:225]: `keypair` stands for an RSA key pair, a dummy one we use for tests.
Now it is in fact an object that has properties for the several 'manifestations' of the keypair:

- keypair.jwk --> the keypair in JWK form
- keypair.jwks --> the keypair in JWKS form (which is the JWK form wrapped in an object with key `keys`)
- keypair.publicKey --> the public key of the keypair
- keypair.privateKey --> the private key of the keypair

Etc etc

Comment (ottokruse) [reply tests/unit/jwt-rsa.test.ts:225]: I  just pushed a commit that adds inline docs to the `keypair` object so that when you hover over it in the IDE, you're explained what it is.

File: src/jwk.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ottokruse) [reply src/jwk.ts]: It's unreffing: https://nodejs.org/api/timers.html#timeoutunref

Moved it now to `node-web-compat` because you were right, it is a specific thing for NodeJS, and all platform specific stuff should be in that file.

**Review by elorzafe — APPROVED** [review]
Awesome work lets ship it 🎉 

Thanks guys! @hakanson @ottokruse @leelalagudu
